upgrade/steps_helpers.go

```diff
@@ -347,16 +346,18 @@ func setUpstreamFromRemoteRepo(
 	return fmt.Errorf("no tag commit that matched '%s' in '%s'", rev, url)
 }
 
-func gitRefsOf(ctx context.Context, url, kind string) (gitRepoRefs, error) {
-	args := []string{"ls-remote", "--" + kind, url}
-	cmd := exec.CommandContext(ctx, "git", args...)
-	out := new(bytes.Buffer)
-	cmd.Stdout = out
-	if err := cmd.Run(); err != nil {
-		return gitRepoRefs{}, fmt.Errorf("git %s: %w", strings.Join(args, " "), err)
-	}
+func gitRefsOf(ctx context.Context, url, kind string) (refs gitRepoRefs, err error) {
+	err = stepv2.PipelineCtx(ctx, "shim", func(ctx context.Context) {
+		refs = gitRefsOfV2(ctx, url, kind)
+	})
+	return
+}
+
+var gitRefsOfV2 = stepv2.Func21("git refs of", func(ctx context.Context, url, kind string) gitRepoRefs {
```

Comment:
Is there a reason we want this to be a var? Do we need to mutate this later?

Reply:
We do not want to mutate it later. It's a `var` because it can't be a `const`.

Reply:
Can it not be a `func`?

Reply:
Not without pain, because we want it to be wrapped in `stepv2.Func21("git refs of", ...)`. The equivalent `func` would be this:

```go
func gitRefsOfV2(ctx context.Context, url, kind string) gitRepoRefs {
	return stepv2.Func21("git refs of", func(ctx context.Context, url, kind string) gitRepoRefs {
		out := stepv2.Cmd(ctx, "git", "ls-remote", "--"+kind, url)

		branchesToRefs := map[string]string{}
		for i, line := range strings.Split(out, "\n") {
			if line == "" {
				continue
			}
			parts := strings.Split(line, "\t")
			contract.Assertf(len(parts) == 2,
				"expected git ls-remote to give '\t' separated values, found line %d: '%s'",
				i, line)
			branchesToRefs[parts[1]] = parts[0]
		}
		return gitRepoRefs{branchesToRefs, kind}
	})(ctx, url, kind)
}
```

That's a lot of code duplication and extra indentation for enforced immutability.

Reply:
I have differing preferences here but hey're 100% non-blocking. It does look annoying to write; let's go with the var implementation.